sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/ReviveManager.java

```diff
@@ -87,44 +43,92 @@ public void run() {
      * Pseudo-code algorithm is this:
      *
      *     // We always start out revived
-     *     List<Requirement> currRequirements = getRequirements();
+     *     List<Requirement> oldRequirements = getRequirements();
      *
      *     while (true) {
-     *         List<Requirement> newRequirements = getRequirements() - currRequirements;
+     *         List<Requirement> currRequirements = getRequirements()
+     *         List<Requirement> newRequirements = oldRequirements - currRequirements;
      *         if (newRequirements.isEmpty()) {
      *             print “No new work”;
      *         } else {
-     *             currRequirements = newRequirements;
+     *             oldRequirements = currRequirements;
      *             revive();
      *         }
      *     }
      *
-     * A natural question is why do we overwrite {@code currRequirements} with {@code newRequirements}?  Anything that
-     * is in {@code currRequirements} has had revive called for it.  Any change with regard to {@code currRequirements}
-     * i.e. {@code newRequirements} implies that those requirements may need an offer which has been declined forever,
-     * so we need to revive.  We cannot maintain a cummulative list in {@code currRequirements} because we may see the
-     * same work twice.
-
-     * e.g.
+     * The invariant maintained is that the oldRequirements have always had revive called for them at least once, giving
+     * them access to offers.
+     *
+     * This case must work:
+     *
      *     kafka-0-broker fails    @ 10:30, it's new work!
      *     kafka-0-broker recovers @ 10:35
      *     kafka-0-broker fails    @ 11:00, it's new work!
      *     ...
      */
-    private void revive() {
-        Set<PodInstanceRequirement> newCandidates = getRequirements(planCoordinator.getCandidates());
+    public void revive(Collection<Step> steps) {
+        Set<WorkItem> currCandidates = getCandidates(steps);
+        Set<WorkItem> newCandidates = new HashSet<>(currCandidates);
         newCandidates.removeAll(candidates);
 
-        logger.debug("Old candidates: {}", candidates);
-        logger.debug("New candidates: {}", newCandidates);
+        logger.info("Candidates, old: {}, current: {}, new:{}", candidates, currCandidates, newCandidates);
+
+        if (!newCandidates.isEmpty()) {
+            if (tokenBucket.tryAcquire()) {
```

Comment:
As I mentioned above, this code path isn't fully exercised by the tests (or at least, I don't see one that does since the defaults of TokenBucket are so high).

I definitely think the code is 100% correct, but I'd feel more comfortable with a test.

Reply:
This code path is exercised by several tests.  Namely every test that successfully acquires a token.

Reply:
Also it's tested extensively by the TokenBucket tests.

Reply:
Also, ReviveManager had total test coverage with the original ReviveManager test except for one line.  I've added a test so that line is now exercised.